man/man1/runbody6.1
.TH RUNBODY6 1NEMO "26 April 2019"
.SH NAME
runbody6 \- Hermite N-body code with optional stellar evolution, NEMO frontend to nbody6 and nbody6++
.SH SYNOPSIS
.nf
\fBrunbody6\fP [parameter=value]
\fBnbody6  \fP\fI < nbody6.in\fP
\fBnbody6++  \fP\fI < nbody6.in\fP
.fi
.SH DESCRIPTION
\fBrunbody6\fP is a NEMO frontend to the Aarseth family NBODY6(++) program.
It creates a run directory, prepares
a parameter input file, and optionally input \fI(snapshot(5NEMO)\fP data, and runs
the NBODY6(++) program. This can be particularly useful if you need to run a lot of
simulations unattended.
.PP
Note that this front-end is not yet designed to run every possible way
in which NBODY6++ can be run, particularly re-runs,
but it most useful to run initial conditions
from a NEMO \fIsnapshot(5NEMO)\fP file.
.PP
Although the \fBkz=\fP vector can be specified, an alternative way
is the indexed parameter access method that \fIgetparam(3NEMO)\fP provides,
which overrides the default \fBkz=\fP. For example to turn off KS, use
.nf
      KZ15=0 KZ30=0
.fi
note the all-caps name of these indexed keywords. Since in NBODY6++ this vector
has 50 elements, overriding them with an indexed parameter may be easier. It also
bypasses any decisions of using the correct default for certain KZ's that is done
in the kz= vector. For example, for nbody6=1 the default value for kz(22)=2 will
be reset to kz(22)=3, but by setting KZ22=2 you can cheat and override these.
.SH PARAMETERS
The following parameters are recognized in any order if the keyword
is also given: 
.TP 20
\fBin=\fP
input file (optional, see \fBnbody=\fP to generate your own input 
data on the fly). If input data are given, it forces KZ(22)=2 (see 
below).  
Default: not used.
.TP
\fBoutdir=\fP
Output run directory. This directory should not exist, although
can be overriden with NEMO's standard \fBerror=\fP system keyword.
No default.
.TP
\fBnbody=\fP
Total particle number. The maximum value (NMAX) is compiled into
the NBODY4 program. Note that even if \fIin=\fP is supplied, this
parameter is needed.
No default.
.TP
\fBnfix=\fP
Output frequency of data save or binaries; KZ(3 & 6)
[Default: 1]
.TP
\fBnrand=\fP
Random number sequence skip.
[Default: 0]
.TP
\fBnrun=\fP
Run identification index.
[Default: 1]
.TP
\fBeta=\fP
Time-step parameter for force polynomials.
[Default: 0.02]
.TP
\fBdeltat=\fP
Output time interval in units of the crossing.
[Default: 0.25]
.TP
\fBtcrit=\fP
Termination time in units of the crossing time.
[Default: 5]
.TP
\fBqe=\fP
Energy tolerance (restart if DE/E > 5*QE & KZ(2) > 1)
[Default: 2e-5]
.TP
\fBeps=\fP
Potential softening parameter     
[Default: 0.01]
.TP
\fBkz=\fP
Non-zero options for alternative paths (up to 50)
.nf
[0 0 1 0 0 0 5 0 0 1  0 0 0 0 2 0 0 0 0 0  1 2 2 0 0 2 0 0 0 2  0 0 2 0 0 0 1 0 0 0  0 0 0 0 0 0 0 0 0 0]
.ta +0.5i
1	COMMON save on unit 1 at end of run (=2: every 100*NMAX steps).
2	COMMON save on unit 2 at output (=1); restart if DE/E > 5*QE (=2).
3	Basic data on unit 3 at output (freq. NFIX; >1: cluster + tail).
4 #	Binary diagnostics on unit 4 (# threshold levels = KZ(4) < 10).
5	Initial conditions (#22 =0; =0: uniform & isotropic sphere;
      	      =1: Plummer; =2: two Plummer models in orbit, extra input;
      	      =3: massive perturber and planetesimal disk, extra input).
      	      =4: massive initial binary, extra input; output on unit 35).
6	Output of significant & regularized binaries (=1, 2, 3 & 4).
7	Lagrangian radii (>0: RSCALE; =2, 3, 4: output units 6, 7, 12;
      	      >=5: density & rms velocity at given radii on unit 26 & 27;
              >=5: average mass at given radii on unit 36 every DELTAT;
      	       =6: Lagrangian radii for two mass groups on unit 31 & 32.
8	Primordial binaries (=1 & >=3; >0: BINOUT; >2: BINDAT; >3: HIDAT).
9	Individual bodies printed at output time (MIN(5**KZ9,NTOT)).
10	Diagnostic KS output (>0: begin; >1: end; >=3: each step).
11 #	Synchronization of circular orbits (suppressed).
12	Disk shocks (=1: standard model) or interstellar clouds (< 0).
13	Scaling of time (1: variable by t_cr; 2: variable by t_r;
      	       -1: constant scaling to t_r; -2: constant scaling to t_c).
14	External force (=1: linearized; -1: cutoff; =2: point-mass galaxy;
      	   =3: point-mass + disk + logarithmic halo in any combination).
15	Triple, quad, chain (#30 > 0) or merger search (>1: full output).
16	Updating of regularization parameters (RMIN, DTMIN & ECLOSE).
17	Modification of ETA (>=1) & ETAU (>1) by tolerance QE.
18	Hierarchical systems (=1: diagnostics; =2: primordial; =3: both).
19	Stellar evolution and mass loss (=1: old supernova scheme;
      	            =3: Eggleton, Tout & Hurley; >4: Chernoff--Weinberg).
20	Initial mass function (=0,1: Salpeter; >1: various, see IMF).
21	Extra output (>0: model, etc; >1: CENTRE; >2: MTRACE; >3: GLOBAL).
22	Initial conditions on unit 10 (=1: output; =2,3(unscaled): input).
23	Escaper removal (>1: diagnostics in file ESC; =2: angles unit #6;
      	                 >1: tidal tails output if #14 = 3).
24	Initial conditions for subsystem (routine SCALE; KZ(24) = #).
25 #	Partial reflection of KS binary orbit (GAMMA < GMIN; suppressed).
25	HR diagnostics of evolving stars (output of B & S on #82 & 83).
26	Slow-down of two-body motion (=1: KS binary; =2: chain binary).
27	Two-body interactions (-2: RADIUS = 0; -1: collision detection;
      	                       =1: sequential circ; > 0: collision).
28	(not used).
29 #	Boundary reflection for hot system (suppressed).
30	Chain regularization (=1: basic; >1: main output; >2: each step).
31	Centre of mass correction after energy check.
32	Increase of output intervals (based on single particle energy).
33	Block-step diagnostics at main output (=2: active pipes).
34 	Roche lobe overflow (suppressed).
35	Time offset (global time from TTOT = TIME + DTOFF; offset = 100).
36	Step reduction for hierarchical systems (not recommended).
37	Step reduction for encounters with high-velocity particles.
38 	Multiple use of GRAPE-6 (sleep 1 sec after each timer check).
39	Neighbour list (=-1: on host; =0: full list or closest on GRAPE).
40      Use force and first deriv to estimate time steps (0), >0 uses Flploy2
41      proto-star evolution properties
42      Initial binary distribution
43      (unused)
44      (unused)
45      (unused)
46      HDF5/binary/ANSI format
47      frequency of KZ(47) output
48      (unused)
49      compute moments of inertia (Chr. Theis)
50      for unpert KS binary.

#	currently surpressed
.fi
.TP
\fBxtpar1=\fP
Mass of external Plummer model (KZ(15) = 1;
[Default: 1]
.TP
\fBxtpar2=\fP
Length scale for Plummer model (KZ(15) = 1)
[Default: 2]
.TP
\fBzmgas=\fP
Mass scale for logarithmic potential (KZ(15) = 2)
.TP
\fBrgas=\fP
Length scale for logarithmic potential (KZ(15) = 2)
.TP
\fBalphas=\fP
Power-law index for initial mass function (routine DATA)
[Default: 2.3]
.TP
\fBbody1=\fP
Maximum particle mass before scaling   
[Default: 5.0]
.TP
\fBbodyn=\fP
Minimum particle mass before scaling   
[Default: 1.0]
.TP
\fBq=\fP
Virial ratio (q=0.5 for virial equilibrium)  
[Default: 0]
.TP
\fBvxrot=\fP
XY-velocity scaling factor (> 0 for solid-body rotation)
[Default: 0]
.TP
\fBvzrot=\fP
Z-velocity scaling factor (not used if VXROT =
[Default: 0]
.TP
\fBrbar=\fP
Virial radius in pc (for scaling to physical
[Default: 1]
.TP
\fBzmbar=\fP
Mean mass in solar units   
[Default: 1]
.TP
\fBxcm=\fP
Displacement for subsystem (routine SCALE; KZ(17)) 
.TP
\fBecc=\fP
Eccentricity of relative motion for subsystem (ECC =< 1)
.TP
\fBKZ#=\fP
This is an indexed keyword, and the upper case version of the kz= array keyword. This
way the user can override single elements of the kz= vector, and using a default
for the long kz= vector. For example "kz=@kz.def KZ22=2 KZ5=2". See \fIgetparam(3NEMO)\fP
for a description of indexed keywords.
.TP
\fBnbody6=0|1\fB
Support mode for nbody6:  1=nbody6, 0=nbody6++ (default)
.TP
\fBexe=\fP
Name of the nbody6 executable in your $PATH (absolute or relative path also allowed).
Both nbody6 and nbody6++ are supported.
The default for \fBexe\fP depends on the value for \fBnbody6=\fP:   1=nbody6  0=nbody6++
Note that for nbody6=1 the default nbody6 executable is the single-processing version,
there is also a parallel version, dubbed \fBnbody7.sse\fP. For nbody6++ there is
the default \fBnbody6++.sse\fP and the possibly faster \fBnbody6++.avx\fP available.
.SH EXAMPLES
\fBnbody6\fP can also be used to generate snapshots. Here is an example
to create a 1024 body homogeneous sphere in the directory
run1 and a plummer sphere in run2.
.nf
    % runbody6 "" run1 1024 tcrit=0 KZ5=0 KZ22=1
    % snapplot run1/OUT3.snap

    % runbody6 "" run2 1024 tcrit=0 KZ5=1 KZ22=1
    % snapplot run2/OUT3.snap

.fi
.PP
You can also supply existing NEMO snapshots as initial conditions, this is actually the best supported way to use
runbody6:
.nf
    % mkplummer p1024 1024
    % runbody6 p1024 run3 tcrit=1
.fi
.SH CAVEATS
Small N systems may need to be tuned with larger \fBrs0=\fP and \fBnnbopt=\fP, but don't even think of trying a 3-body problem.
.PP
The dreaded CALCULATIONS HALTED message means you will need to consult the NBODY6++ manual
.PP
The interface has been primarely tested with nbody6++, it's possible that some differences between
nbody6 and nbody6++ have not been taken care of.
.SH BUGS
In mode nbody6=1 the end-time does not seem to be reached. Seem to halt at tcrit-dtadj ?
.SH SEE ALSO
runbody1(1NEMO), nbody2(1NEMO), nbody4(5NEMO), snapshot(5NEMO), u3tos(1NEMO), stou4(1NEMO), snapprint(1NEMO)
.PP
.nf
nbody6:    ftp://ftp.ast.cam.ac.uk/pub/sverre/nbody6/nbody6.tar.gz
nbody6++:  https://github.com/nbodyx/Nbody6ppGPU
.fi
.SH FILES
.nf
.ta +2i
$NEMO/usr/aarseth	support to install and test the codes
$outdir/OUT3.snap	particle dump (see \fIu3tos(1NEMO)\fP) in \fIsnapshot(5NEMO)\fP format.
.fi
.SH AUTHOR
Peter Teuben
.SH UPDATE HISTORY
.nf
.ta +1.0i +4.0i
20-feb-2019	V0.1 cloned off runbody4, some of keyword descriptions not complete yet		PJT
22-feb-2019	V0.5 support nbody6/nbody6++	PJT
5-apr-2019	V0.7 support writing snapshots without virtual particles	PJT
26-apr-2019	V0.8 reversed logic of nbody6=, fix  KZ22 option for nbody6=1	PJT
.fi


src/nbody/evolve/aarseth/tools/runbody6.c
/*
 *  MAIN.C: NEMO driver for nbody6
 *	
 *  19-feb-2019  V0.1     draft for nbody6++                             PJT
 *                         
 *  CAVEAT:    this is optimized to read a NEMO snapshot and process that and
 *             write a NEMO snapshot.
 *             Although it (should) work with nbody6, it was developed with nbody6++
 * 
 */

#include <stdinc.h>
#include <getparam.h>
#include <vectmath.h>
#include <filestruct.h>
#include <history.h>
#include <run.h>

#include <snapshot/snapshot.h>
#include <snapshot/body.h>
#include <snapshot/get_snap.c>

#include <sys/types.h>
#include <fcntl.h>
#include <unistd.h>

string defv[] = {
    "in=\n            input file (optional - see nbody= ) ",
    "outdir=???\n     output run directory (required)",

    "nbody=1000\n     Total particle number (<= NMAX) if no in= given",
    "nfix=1\n         Output frequency of data save or binaries; KZ(3 & 6)",
    "ncrit=5\n        Final particle number (alternative termination criterion)",
    "nrand=123\n      Random number seed",
    "nnbopt=20\n      Desired optimal neighbor number (NNBMAX in nbody6)",
    "nrun=1\n         Run identification index",
    "ncomm=10\n       Frequency to store dumping data (nbody6++ only)",

    "etai=0.02\n      Time-step parameter for irregular force polynomial",
    "etar=0.02\n      Time-step parameter for regular force polynomial",
    "rs0=0.1\n        Initial guess for all radii of neighbor spheres",
    "dtadj=$deltat\n  Time interval for parameter adjustment and energy check",
    "deltat=0.25\n    Output time interval",
    "tcrit=10.0\n     Termination time",
    "qe=2e-5\n        Energy tolerance (restart if DE/E > 5*QE & KZ(2) > 1)",
    "rbar=1.0\n       mean radius of system",
    "zmbar=0.5\n      mean mass of system, in solar units",

    "kz=0 0 1 0 0 1 5 0 0 1  0 0 0 0 2 1 1 0 0 0  1 2 0 0 0 2 0 0 0 2  0 0 2 0 0 0 1 0 0 1  0 0 0 0 0 0 0 0 0 0\n",
      "Non-zero options for alternative paths (see below)\n"
      "       1  COMMON save on unit 1 at end of run (=2: every 100*NMAX steps).\n"
      "       2  COMMON save on unit 2 at output (=1); restart if DE/E > 5*QE (=2).\n"
      "       3  Basic data on unit 3 at output (freq. NFIX; >1: cluster + tail).\n"
      "     # 4  Binary diagnostics on unit 4 (# threshold levels = KZ(4) < 10).\n"
      "       5  Initial conditions (#22 =0; =0: uniform & isotropic sphere;\n"
      "                =1: Plummer; =2: two Plummer models in orbit, extra input;\n"
      "                =3: massive perturber and planetesimal disk, extra input).\n"
      "                =4: massive initial binary, extra input; output on unit 35).\n"
      "       6  Output of significant & regularized binaries (=1, 2, 3 & 4).\n"
      "       7  Lagrangian radii (>0: RSCALE; =2, 3, 4: output units 6, 7, 12;\n"
      "                >=5: density & rms velocity at given radii on unit 26 & 27;\n"
      "                 =6: Lagrangian radii for two mass groups on unit 31 & 32.\n"
      "       8  Primordial binaries (=1 & >=3; >0: BINOUT; >2: BINDAT; >3: HIDAT).\n"
      "       9  Individual bodies printed at output time (MIN(5**KZ9,NTOT)).\n"
      "      10  Diagnostic KS output (>0: begin; >1: end; >=3: each step).\n"
      "    # 11  Synchronization of circular orbits (suppressed).\n"
      "      12  Disk shocks (=1: standard model) or interstellar clouds (< 0).\n"
      "      13  Scaling of time (1: variable by t_cr; 2: variable by t_r;\n"
      "                 -1: constant scaling to t_r; -2: constant scaling to t_c).\n"
      "      14  External force (=1: linearized; -1: cutoff; =2: point-mass galaxy;\n"
      "             =3: point-mass + disk + logarithmic halo in any combination).\n"
      "      15  Triple, quad, chain (#30 > 0) or merger search (>1: full output).\n"
      "      16  Updating of regularization parameters (RMIN, DTMIN & ECLOSE).\n"
      "      17  Modification of ETA (>=1) & ETAU (>1) by tolerance QE.\n"
      "      18  Hierarchical systems (=1: diagnostics; =2: primordial; =3: both).\n"
      "      19  Stellar evolution and mass loss (=1: old supernova scheme;\n"
      "                      =3: Eggleton, Tout & Hurley; >4: Chernoff--Weinberg).\n"
      "      20  Initial mass function (=0,1: Salpeter; >1: various, see IMF).\n"
      "      21  Extra output (>0: model, etc; >1: CENTER; >2: MTRACE; >3: GLOBAL).\n"
      "      22  Initial conditions on unit 10 (=1: output; =2,3(unscaled): input).\n"
      "      23  Escaper removal (>1: diagnostics in file ESC; =2: angles unit #6;\n"
      "                           >1: tidal tails output if #14 = 3).\n"
      "      24  Initial conditions for subsystem (routine SCALE; KZ(24) = #).\n"
      "    # 25  Partial reflection of KS binary orbit (GAMMA < GMIN; suppressed).\n"
      "      25  HR diagnostics of evolving stars (output of B & S on #82 & 83).\n"
      "      26  Slow-down of two-body motion (=1: KS binary; =2: chain binary).\n"
      "      27  Two-body interactions (-2: RADIUS = 0; -1: collision detection;\n"
      "                                 =1: sequential circ; > 0: collision).\n"
      "      28  (not used).\n"
      "    # 29  Boundary reflection for hot system (suppressed).\n"
      "      30  Chain regularization (=1: basic; >1: main output; >2: each step).\n"
      "      31  Center of mass correction after energy check.\n"
      "      32  Increase of output intervals (based on single particle energy).\n"
      "      33  Block-step diagnostics at main output (=2: active pipes).\n"
      "    # 34  Roche lobe overflow (suppressed).\n"
      "      35  Time offset (global time from TTOT = TIME + DTOFF; offset = 100).\n"
      "      36  Step reduction for hierarchical systems (not recommended).\n"
      "      37  Step reduction for encounters with high-velocity particles.\n"
      "    # 38  Multiple use of GRAPE-6 (sleep 1 sec after each timer check).\n"
      "      39  Neighbour list (=-1: on host; =0: full list or closest on GRAPE).\n"
      "      40  (not used).\n"
      "    # 41..50  TBD",
    

    "dtmin=1e-4\n     time-step criterion for regularization search",
    "rmin=1e-3\n      distance criterion for regularization search",
    "etau=0.1\n       ",
    "eclose=1.0\n     ",
    "gmin=1e-6\n      ",
    "gmax=0.001\n     ",
    "smax=0.25\n      ",

    "alpha=2.3\n      Power-law index for initial mass function",
    "body1=10.0\n     Maximum particle mass before scaling",
    "bodyn=0.2\n      Minimum particle mass before scaling",
    "nbin0=0\n        number of primordial binaries",
    "nhi0=0\n         number of primordial hierarchical binaries",
    "zmet=0.001\n     Metal abundance (0.03..0.0001)",
    "epoch0=0\n       epoch of star formation (in Myr)",
    "dtplot=10.0\n    plotting interval (nbody units)",

    /* some kz(5) > 0 options here: APO,ECC,N2,SCALE SEMI, M1,M2,ZMH,RCUT */

    "q=0.5\n          Virial ratio (q=0.5 for virial equilibrium)",
    "vxrot=0\n        XY velocity scaling",
    "vzrot=0\n        Z velocity scaling",
    "rtide=1\n        Unscaled tidal radius for kz(14)=2 and kz(22) >= 2",

    /* some more:   GMG,RG0 if kz(14)=2 or 3*/

    /* ah, here they are */

    "apo=6.0\n        kz(5) = 2 or 3",
    "ecc=0.5\n        Eccentricity of relative motion for subsystem (ECC =< 1)",
    "n2=500\n         kz(5) = 2",
    "scale=0.5\n      kz(5) = 2 or 3",
    "dmin=3\n         kz(5) = 3",
    "semi=1e-3\n      ",
    "m1=25.0\n        ",
    "m2=25.0\n        ",
    "ratio=0.0\n      ",
    "range=100.0\n    ",
    
    "kstart=1\n       Running mode (1=new 2=restart 3,4,5=restart w/ new par",
    "tcomp=10.0\n     Maximum allowed running time (minutes) **ignored **",
    "gpid=0\n         Number of grape boards (starting from 0) ** ignored **",

    "format=%g\n      Format used for fort.10 input conditions if in= used",
    "KZ#=\n           [indexed] Override some kz= keywords",
    "nbody6=0\n       run mode : 1=nbody6  0=nbody6++",
    "exe=\n           Name of the executable (default see nbody6=)",

    "VERSION=0.8b\n   13-may-2019 PJT",
    NULL,
};

string usage="NEMO frontend to the NBODY6(++) code";

string cvsid="$Id$";

#define KZ_MAX  50

void nemo_main(void)
{
  int nbody, nfix, ncrit, nrand, nnbopt, nrun, ncomm, kstart, gpid, nbin0, nhi0, n2;
  real etai, etar, rs0, dtadj, deltat, tcrit, qe, eps, tcomp;
  int k, nkz, kz[KZ_MAX], KZ[KZ_MAX];
  int maxidx, bits;
  real dtmin,rmin,etau,eclose,gmin,gmax,smax,tsnap;
  real alpha, body1, bodyn, zmet, epoch0, dtplot;
  real q, rbar, zmbar;
  real apo, ecc, dmin, semi, scale, m1, m2, ratio, range;
  Body *bp, *btab = NULL;

  string exefile = getparam("exe");
  string parfile = "nbody6.in";
  string rundir = getparam("outdir");
  string fmt = getparam("format");
  int nbody6_mode = getiparam("nbody6");
  string infile;
  char dname[256], runcmd[256], fmt7[256];
  stream datstr, histr, instr, outstr;

  warning("This is a not fully tested pre-release version %s",getparam("VERSION"));
  if (!hasvalue("exe")) {
    exefile = (nbody6_mode == 1) ? "nbody6" : "nbody6++";
    dprintf(0,"exefile=%s\n",exefile);
  }

  kstart = getiparam("kstart");
  tcomp =  getdparam("tcomp");
  gpid = getdparam("gpid");
  
  nfix = getiparam("nfix");
  ncrit = getiparam("ncrit");
  nrand = getiparam("nrand");
  nnbopt = getiparam("nnbopt");
  nrun = getiparam("nrun");
  ncomm = getiparam("ncomm");
  
  etai = getdparam("etai");
  etar = getdparam("etar");
  rs0 = getdparam("rs0");
  dtadj = getdparam("dtadj");
  deltat = getdparam("deltat");
  tcrit = getdparam("tcrit");
  qe = getdparam("qe");
  rbar =  getdparam("rbar");
  zmbar =  getdparam("zmbar");
  
  nkz = nemoinpi(getparam("kz"),kz,KZ_MAX);
  if (nkz < 0) error("%d Syntax error kz=%s",nkz,getparam("kz"));
  for (k=nkz; k<KZ_MAX; k++) kz[k]=0;
  for (k=0; k<KZ_MAX; k++) dprintf(1,"%d ",kz[k]);
  dprintf(1,"\n");

  if (nbody6_mode == 1) {
    if (kz[21] == 2) kz[21] = 3;  /* prevent scaling in nbody6 */
  }

  for (k=0; k<KZ_MAX; k++) {
    if (indexparam("KZ",k+1)) {
      dprintf(0,"KZ %d=%d\n",k+1,getiparam_idx("KZ",k+1));
      kz[k] = getiparam_idx("KZ",k+1);
    }
  }

  if (hasvalue("in")) {
    /* read snapshot and grab nbody.... */

    instr = stropen(getparam("in"),"r");
    get_history(instr);
    if (!get_tag_ok(instr, SnapShotTag)) error("no snapshot");
    get_snap(instr, &btab, &nbody, &tsnap, &bits);
    strclose(instr);
    if (kz[21] == 0) {
      warning("patching kz[22]=2 since in= was given");
      if (nbody6_mode == 0)
	kz[21] = 2;          // nbody6++
      else
	kz[21] = 3;          // nbody6
    }
  } else
    nbody = getiparam("nbody");


  dtmin = getdparam("dtmin");
  rmin = getdparam("rmin");
  etau = getdparam("etau");
  eclose= getdparam("eclose");
  gmin = getdparam("gmin");
  gmax = getdparam("gmax");
  smax = getdparam("smax");  

  
  alpha = getdparam("alpha");
  body1 = getdparam("body1");
  bodyn = getdparam("bodyn");
  nbin0 = getiparam("nbin0");
  nhi0 = getiparam("nhi0");
  zmet = getdparam("zmet");
  epoch0 = getiparam("epoch0");
  dtplot = getiparam("dtplot");

  q = getdparam("q");

  /* optional things, depending on kz5 */
  apo = getdparam("apo");
  ecc = getdparam("ecc");
  n2 = getiparam("n2");
  dmin = getdparam("dmin");
  scale = getdparam("scale");
  semi = getdparam("semi");
  m1 = getdparam("m1");
  m2 = getdparam("m2");
  ratio = getdparam("ratio");
  range = getdparam("range");

  /* there are more optional kz-dependant keywords...
     see nbody4.5, they have not been encoded here
     by lack of Cambridge time 
  */



  run_mkdir(rundir);

  sprintf(dname,"%s/%s",rundir,parfile);
  datstr = stropen(dname,"w");    
  
    /*  New Run */
  
  if (kstart == 1) {

    if (nbody6_mode == 0)
      fprintf(datstr,"%d 999999.9 1.E6 40 40 640\n",kstart);
    else
      fprintf(datstr,"%d 999999.9\n",kstart);

    if (nbody6_mode == 0)
      fprintf(datstr,"%d %d %d %d %d %d %d\n",nbody,nfix,ncrit,nrand,nnbopt,nrun,ncomm);
    else
      fprintf(datstr,"%d %d %d %d %d %d\n",nbody,nfix,ncrit,nrand,nnbopt,nrun);

    if (nbody6_mode == 0)
      fprintf(datstr,"%g %g %g %g %g %g %g %g %g\n",etai,etar,rs0,dtadj,deltat,tcrit,qe,rbar,zmbar);
    else
      fprintf(datstr,"%g %g %g %g %g %g %g %g %g\n",etai,etar,rs0,dtadj,deltat,tcrit+dtadj,qe,rbar,zmbar);   // work around tcrit bug in nbody6

    for (k=0; k<KZ_MAX; k++) {
      fprintf(datstr,"%d ",kz[k]);
      if (k%10 == 9)     fprintf(datstr,"\n");
    }
    if (nbody6_mode == 0)
      fprintf(datstr,"%g %g %g %g %g %g %g\n",dtmin,rmin,etau,eclose,gmin,gmax,smax);
    else
      fprintf(datstr,"%g %g %g %g %g %g\n",dtmin,rmin,etau,eclose,gmin,gmax);
    
    fprintf(datstr,"%g %g %g %d %d %g %g %g\n",alpha,body1,bodyn,nbin0,nhi0,zmet,epoch0,dtplot);

    if (kz[4] == 2)
      fprintf(datstr,"%g %g %d %g\n",apo,ecc,n2,scale);
    else if (kz[4] == 3) 
      fprintf(datstr,"%g %g %g %g\n",apo,ecc,dmin,scale);
    else if (kz[4] == 4) 
      fprintf(datstr,"%g %g %g %g\n",semi,ecc,m1,m2);

    if (nbody6_mode == 0)
      fprintf(datstr,"%g 0  0 0 \n",q); 
    else
      fprintf(datstr,"%g 0.0 0.0 0.0 %g\n",q,smax);    // READ (5,*)  Q, VXROT, VZROT, RTIDE, SMAX

    if (kz[7] == 1 || kz[7] == 3)
      fprintf(datstr,"%g %g %g %g 0 0 0 \n",semi,ecc,ratio,range);

    /*  this one is not well documented - gleaned by Massimo from source code */
    fprintf(datstr,"20000.0 2 0\n");
    fprintf(datstr,"20000.0 2 0\n");
    fprintf(datstr,"# this last bogus line should not bother nbody6(++) as it's not supposed to be read\n");
    fprintf(datstr,"# other perhaps useful info: nbody6=%d exefile=%s\n",nbody6_mode,exefile);

    strclose(datstr);

    run_cd(rundir);

    histr = stropen("history","w");
    put_history(histr);
    strclose(histr);

    if (hasvalue("in")) {
      dprintf(1,"Using ascii printout with format=%s to convert data for nbody6\n",fmt);
      sprintf(fmt7,"%s %s %s %s %s %s %s\n",fmt,fmt,fmt,fmt,fmt,fmt,fmt);
      if (nbody6_mode == 0)
	outstr = stropen("dat.10","w");
      else
	outstr = stropen("fort.10","w");
      for (bp=btab; bp<btab+nbody; bp++)
	fprintf(outstr,fmt7,
		Mass(bp),
		Pos(bp)[0],Pos(bp)[1],Pos(bp)[2],
		Vel(bp)[0],Vel(bp)[1],Vel(bp)[2]);
      strclose(outstr);
    }

    sprintf(runcmd,"%s < %s",exefile,parfile);
    run_sh(runcmd);

    if (nbody6_mode == 0)
      sprintf(runcmd,"cat `ls -tr conf.3_*` > OUT3; u3tos OUT3 OUT3.snap mode=6 nbody=%d ; rm OUT3",nbody);
    else
      sprintf(runcmd,"u3tos in=OUT3 out=OUT3.snap mode=1 nbody=%d",nbody);
    run_sh(runcmd);    
    
  } else {
    error("kstart=%d not supported for NBODY6(++) yet; you will need to do it manually",kstart);
    // for nbody6_mode==0
    // kstart=2:   no further lines needed
    //        3:   one line: DTADJ,DELTAT,DTADJ,TNEXT,TCRIT,QE,J,K
    //        4:   one line: ETAI,ETAR,ETAU,DTMIN,RMIN,NCRIT,NNBOPT,SMAX
    //        5:
    // for nbody6_mode==1
    // kstart=2:   standard restart
    //       >2:   read new parameters (see modify.f)
  }
}
